Assets/UI/worldinput_CQUI.lua

```diff
@@ -0,0 +1,559 @@
+    if (uiKey == Keys.S) then
+      if (formationClass == "FORMATION_CLASS_AIR") then
+        -- M4A: Not certain that this command actually is functional
+        UnitManager.RequestCommand(UI.GetHeadSelectedUnit(), UnitOperationTypes.AIR_ATTACK);
+        cquiHandledKey = true;
+      end
+    end
+  end -- Classic binds that overlap with enhanced binds
+
+    -- Focus Capital hotkey
+  if (uiKey == Keys.VK_HOME) then
+    UI.SelectCity(Players[Game.GetLocalPlayer()]:GetCities():GetCapitalCity());
+    cquiHandledKey = true;
+  end
+
+  if (uiKey == Keys.VK_BACK) then
+    if (unitType ~= nil) then
+      UnitManager.RequestCommand(UI.GetHeadSelectedUnit(), UnitCommandTypes.CANCEL);
+      cquiHandledKey = true;
+    end
+  end
+
+  if (uiKey == Keys.C and CQUI_isAltDown == true) then
+    if (unitType == "UNIT_BUILDER") or (unitType == "UNIT_MILITARY_ENGINEER") then
+      UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), UnitOperationTypes.HARVEST_RESOURCE);
+      UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), UnitOperationTypes.REMOVE_FEATURE);
+      -- Use false as we need to let the unmodifed script deal with the Alt Key
+      cquiHandledKey = false;
+    end
+  end
+
+  if (uiKey == Keys.F) then
+    if (unitType == "UNIT_BUILDER") then
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_FISHING_BOATS"].Hash);
+      cquiHandledKey = true;
+    elseif (unitType == "UNIT_MILITARY_ENGINEER") then
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_FORT"].Hash);
+      cquiHandledKey = true;
+    end
+  end
+
+  if (uiKey == Keys.H and unitType == "UNIT_BUILDER") then
+    CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_CAMP"].Hash);
+    cquiHandledKey = true;
+  end
+
+  if (uiKey == Keys.I and unitType == "UNIT_BUILDER") then
+    CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_FARM"].Hash);
+    cquiHandledKey = true;
+  end
+
+  if (uiKey == Keys.L and unitType == "UNIT_BUILDER") then
+    CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_LUMBER_MILL"].Hash);
+    cquiHandledKey = true;
+  end
+
+  if (uiKey == Keys.N) then
+    cquiHandledKey = true;
+    if (unitType == "UNIT_BUILDER") then
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_MINE"].Hash);
+    else
+      UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), UnitOperationTypes.WMD_STRIKE);
+    end
+  end
+
+  if (uiKey == Keys.O) then
+    if (unitType == "UNIT_BUILDER") then
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_OIL_WELL"].Hash);
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_OFFSHORE_OIL_RIG"].Hash);
+      cquiHandledKey = true;
+    end
+  end
+
+  if (uiKey == Keys.P) then
+    if (CQUI_isShiftDown) then
+      CQUI_isShiftDown = false;
```

Comment:
The clearing of the shift key flag is from old code, but do you know if it serves a purpose? I was thinking that this is the only reason why the clearing of the shift and alt key flags had to be placed in two places. So I was just curious if you knew why.

Reply:
not sure, actually.  looking at it, this seems like if you kept shift down it would only respond to the first.

and the more I look at it, I'm not positive I need to do cquiHandledKey=false on any of the keys involving the Alt.  When Alt is the KeyUp, then it'd be handled.

Reply:
removed it there, seems to work ok.